smartcontract/service/native/utils/linked_list.go

```diff
@@ -0,0 +1,385 @@
+
+		node, err = makeLinkedlistNode(qhead.next, item, qhead.payload)
+		if err != nil {
+			//log.Trace(err)
+			return err
+		}
+		PutBytes(native, append(index, head...), node) //head.next = head.next, head.prev = item,
+		// head.payload = head.payload
+		PutBytes(native, index, item) // item becomes head
+	}
+	return nil
+}
+
+func LinkedlistDelete(native *native.NativeService, index []byte, item []byte) (bool, error) {
+	null := []byte{}
+	if item == nil {
+		return false, errors.NewErr("[linked list] invalid item")
+	}
+	q, err := getListNode(native, index, item)
+	if err != nil {
+		return false, err
+	}
+	if q == nil {
+		return false, nil
+	}
+
+	prev, next := q.prev, q.next
+	if prev == nil {
+		if next == nil {
+			PutBytes(native, index, null) //clear linked list
+		} else {
+			qnext, err := getListNode(native, index, next)
+			if err != nil {
+				return false, err
+			}
+			node, err := makeLinkedlistNode(qnext.next, null, qnext.payload) //qnext.next = qnext.next
+			if err != nil {                                                  // qnext.prev = nil
+				return false, err
+			}
+			PutBytes(native, append(index, next...), node)
+			PutBytes(native, index, next) //next becomes head
+		}
+	} else {
+		if next == nil {
+			qprev, err := getListNode(native, index, prev)
+			if err != nil {
+				return false, err
+			}
+			node, err := makeLinkedlistNode(null, qprev.prev, qprev.payload) //qprev becomes end
+			if err != nil {
+				return false, err
+			}
+			PutBytes(native, append(index, prev...), node)
+		} else {
+			qprev, err := getListNode(native, index, prev)
+			if err != nil {
+				return false, err
+			}
+			qnext, err := getListNode(native, index, next)
+			if err != nil {
+				return false, err
+			}
+			node_prev, err := makeLinkedlistNode(next, qprev.prev, qprev.payload) //
+			if err != nil {
+				return false, err
+			}
+			node_next, err := makeLinkedlistNode(qnext.next, prev, qnext.payload)
+			if err != nil {
+				return false, err
+			}
+			PutBytes(native, append(index, prev...), node_prev)
+			PutBytes(native, append(index, next...), node_next)
+		}
+	}
+	PutBytes(native, append(index, item...), null)
+	return true, nil
+}
+
+func LinkedlistGetItem(native *native.NativeService, index []byte, item []byte) (*LinkedlistNode, error) {
+	if item == nil {
+		return nil, errors.NewErr("[linkedlist getNext] item is nil")
+	}
+	q, err := getListNode(native, index, item)
+	if err != nil {
+		return nil, err
+	}
+	return q, nil
+}
+
+func LinkedlistGetHead(native *native.NativeService, index []byte) ([]byte, error) {
+	head, err := getListHead(native, index)
+	if err != nil {
+		return nil, err
+	}
+	return head, nil
+}
+func LinkedlistGetNumOfItems(native *native.NativeService, index []byte) (int, error) {
+	n := 0
+	head, err := getListHead(native, index)
+	if err != nil {
+		return 0, err
+	}
+	q := head
+	for q != nil {
+		n += 1
+		qnode, err := getListNode(native, index, q)
+		if err != nil {
+			return 0, err
+		}
```

Comment:
iterating all of the list node is too expensive, you may need store this meta info in the list's `head` key.